ldap/servers/slapd/backend.c

```diff
@@ -204,6 +206,108 @@ slapi_be_gettype(Slapi_Backend *be)
+
+            /* ignore special backends */
+            if ((strcmp(suffix, "cn=schema") == 0) ||
+                (strcmp(suffix, "cn=config") == 0)) {
+                goto next;
+            }
+
+            /* search for ("smart") referral entries */
+            search_pb = slapi_pblock_new();
+            server_ctrls = (LDAPControl **)slapi_ch_calloc(2, sizeof(LDAPControl *));
+            server_ctrls[0] = (LDAPControl *)slapi_ch_malloc(sizeof(LDAPControl));
+            server_ctrls[0]->ldctl_oid = slapi_ch_strdup(LDAP_CONTROL_MANAGEDSAIT);
+            server_ctrls[0]->ldctl_value.bv_val = NULL;
+            server_ctrls[0]->ldctl_value.bv_len = 0;
+            server_ctrls[0]->ldctl_iscritical = '\0';
+            slapi_search_internal_set_pb(search_pb, suffix, LDAP_SCOPE_SUBTREE,
+                                         filter, NULL, 0, server_ctrls, NULL,
+                                         (void *) plugin_get_default_component_id(), 0);
+            slapi_search_internal_pb(search_pb);
+            slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
+            if (LDAP_SUCCESS != rc) { /* plugin is not available */
+                exist_referral = 1; /* be safe, assume there is a referral somewhere */
+            }
+
+            slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_SEARCH_ENTRIES, &entries);
+            slapi_pblock_get(search_pb, SLAPI_PLUGIN_INTOP_SEARCH_REFERRALS, &referrals);
+            if ((entries && entries[0]) || referrals) {
+                exist_referral = 1;
+            }
+            slapi_free_search_results_internal(search_pb);
+            slapi_pblock_destroy(search_pb);
+        }
+
+next:
+        /* next backend */
+        be = slapi_get_next_backend(cookie);
+    }
+    slapi_ch_free((void **) &cookie);
+
+    /* now toggle the flag if the status changed */
+    if (exist_referral) {
+        if (exist_referrals == 0) {
+            /* it existed no referral but now it exists at least a new referral entry */
+            slapi_log_err(SLAPI_LOG_INFO, "slapd_daemon", "New referral entries are detected (returned to SRCH req)\n");
+            exist_referrals = 1;
+        }
+    } else if (exist_referrals) {
+        /* now there is no more referral */
+        slapi_log_err(SLAPI_LOG_INFO, "slapd_daemon", "No more referral entries\n");
+        exist_referrals = 0;
+    }
+}
+
+/* schedule periodic call to referral_check */
+void
+slapi_referral_check_init(void)
+{
+    referral_check_ctx = slapi_eq_repeat_rel(referral_check,
+                                             NULL, (time_t)0,
+                                             /* 30 sec in milliseconds */
+                                             30 * 1000);
+}
+void
+slapi_referral_check_stop(void)
+{
+    if (referral_check_ctx) {
+        slapi_eq_cancel_rel(referral_check_ctx);
+        referral_check_ctx = NULL;
+    }
+}
+int
+slapi_exist_referrals(void)
+{
+    return exist_referrals;
```

Comment:
Do we need a lock here?   I also recall that "exist_referrals" should be a static uint64_t in this case (something to do safety but I don't remember the details anymore)

Reply:
Thanks for the review Mark. I think it does not need a lock. The update is periodic so if the caller reads an outdated value it is not a big deal.
Regarding the size, it looks a good point. Indeed the cache line will be in competition between workers, I will do test with cache line alignment.

Reply:
Performance wise, there is no benefit when making it uint64_t and aligning it on cache size

Reply:
It's probably not needing a lock, but an atomic to set it with RELEASE and to check the exist with ACQUIRE is correct.

Reply:
Although using lock or atomic operation as described by @Firstyear is the standard way to set and check flags in different threads, in this specific case I agree with Thierry: it is an overkill.
We do not really care of the timing (as @Firstyear said in another comment we could check if the flag  once every few minutes (so if the change stay in the cpu pipe line until next context switch, does not really matter)